### PR TITLE
Grid fix

### PR DESCRIPTION
**Pull request by lauragustafson**
Fixes a number of bugs found in Tuner/Recommender when creating evaluation scripts.

### CODE REVIEW

File: btb/tuning/gcp.py [hunk 1]
```diff
@@ -125,7 +127,12 @@ def jitter(x, range):
             return
 
         # -- Non-parametric model of 'y', estimated with kernel density
-        kernel_pdf = st.gaussian_kde(y)
+        try:
+            kernel_pdf = st.gaussian_kde(y)
+        except LinAlgError as e:
```
Comment (micahjsmith) [inline btb/tuning/gcp.py]: maybe log this? I worry that LinAlgErrors could result from other conditions besides singular matrices and these bugs would be totally lost. Log on `debug` with manually printed stacktrace or log using `exception` level with automatically printed stacktrace, depending on your judgment?

File: btb/tuning/gcp.py [hunk 2]
```diff
@@ -3,11 +3,13 @@
 import logging
 
 import numpy as np
+from numpy.linalg import LinAlgError
 import scipy.stats as st
 from scipy.stats import norm
 from sklearn.gaussian_process import GaussianProcessRegressor
 
 from btb.tuning.tuner import BaseTuner
+from btb.tuning.gp import GP
```
Comment (micahjsmith) [inline btb/tuning/gcp.py]: sort imports

File: btb/tuning/gcp.py [hunk 3]
```diff
@@ -141,9 +148,14 @@ def jitter(x, range):
...
             if self.tunables[ki][1].is_integer:
                 columnX = jitter(columnX, self.tunables[ki][1].range)
-            kernel_pdf = st.gaussian_kde(columnX)
-            kernel_cdf = make_cdf(kernel_pdf)
-            kernel_ppf = make_ppf(kernel_pdf)
+            try:
+                kernel_pdf = st.gaussian_kde(columnX)
+                kernel_cdf = make_cdf(kernel_pdf)
+                kernel_ppf = make_ppf(kernel_pdf)
+            except LinAlgError as e:
```
Comment (micahjsmith) [inline btb/tuning/gcp.py]: same as above

File: btb/tuning/gcp.py [hunk 4]
```diff
@@ -171,6 +186,15 @@ def get_valid_row(U):
             V = np.copy(U[ind_OK, :])
             return V, ind_OK
 
+        # Check if GCP failed to fit due to singular matrix
+        if self.gcp is None:
+            print(
```
Comment (micahjsmith) [inline btb/tuning/gcp.py]: don't print anything, use logger instead at appropriate log level

File: btb/tuning/gcp.py [hunk 5]
```diff
@@ -233,6 +257,8 @@ def _acquire(self, predictions):
                 the gaussian process generated for a series of
                 proposed hyperparameters.
         """
+        if self.X.shape[0] < self.r_minimum:
+            return super(GCPEi, self)._acquire(predictions)
```
Comment (micahjsmith) [inline btb/tuning/gcp.py]: nit: this logically is an if/else block,  consider making it one explicitly

File: tests/btb/recommendation/test_recommender.py [hunk 6]
```diff
@@ -36,8 +37,9 @@ def test_fit(self):
 
     def test__get_candidates_all(self):
         recommender = BaseRecommender(self.dpp_matrix)
-        X = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
-        recommender.dpp_vector = X
+        #X = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
+        #recommender.dpp_vector = X
+        recommender.tried_pipelines = set([])
```
Comment (micahjsmith) [inline tests/btb/recommendation/test_recommender.py]: nit: `set()`

File: tests/btb/recommendation/test_recommender.py [hunk 7]
```diff
@@ -49,19 +51,34 @@ def test__get_candidates_all(self):
 
     def test__get_candidates_some(self):
         recommender = BaseRecommender(self.dpp_matrix)
-        X = np.array([0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0])
-        recommender.dpp_vector = X
+        #X = np.array([0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0])
```
Comment (micahjsmith) [inline tests/btb/recommendation/test_recommender.py]: delete commented out lines

File: tests/btb/recommendation/test_recommender.py [hunk 8]
```diff
@@ -49,19 +51,34 @@ def test__get_candidates_all(self):
...
+            expected,
+        )
+        assert set([0, 2]) == recommender.tried_pipelines
+
     def test__get_candidates_none(self):
         recommender = BaseRecommender(self.dpp_matrix)
-        X = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
-        recommender.dpp_vector = X
+        recommender.tried_pipelines = set(range(16))
+        #X = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
```
Comment (micahjsmith) [inline tests/btb/recommendation/test_recommender.py]: delete

File: tests/btb/recommendation/test_recommender.py [hunk 9]
```diff
@@ -49,19 +51,34 @@ def test__get_candidates_all(self):
...
+        recommender.add({0: 0, 2: 0})
+        expected = np.array(
+            [1, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
+        )
+        candidates = recommender._get_candidates()
+        np.testing.assert_array_equal(
+            candidates,
+            expected,
+        )
+        assert set([0, 2]) == recommender.tried_pipelines
```
Comment (micahjsmith) [inline tests/btb/recommendation/test_recommender.py]: always prefer the unittest.TestCase assertion methods because they provide more debugging information and flexibility
```python
expected_tried_pipelines = set([0, 2])
self.assertEqual(expected_tried_pipelines, recommender.tried_pipelines)
```

See https://docs.python.org/3/library/unittest.html#assert-methods for more details

File: tests/btb/tuning/test_gp.py [hunk 10]
```diff
@@ -118,7 +118,7 @@ def test_predict_x_lt_r_min(self, uniform_mock):
         predicted = tuner.predict(X)
 
         # assert
-        expected = np.array([0.8, 0.9])
+        expected = np.array([[0.8, 0], [0.9, 0]])
```
Comment (micahjsmith) [inline tests/btb/tuning/test_gp.py]: nit: Since you have a first column of values and a second column of 0s, it might make sense to type it as
```python
np.array([[0.8, 0.9], [0.0, 0.0]]).T
```
Especially if you expanded the tests, it would be clearer and less error-prone to add new rows

File: tests/btb/tuning/test_gp.py [hunk 11]
```diff
@@ -186,13 +186,33 @@ class TestGPEi(TestCase):
     # METHOD: _acquire(self, predictions)
     # VALIDATE:
     #     * return values according to the formula
+    #@patch('btb.tuning.gp.GP')
+    def test__acquire_less_than_rmin(self): #, gp_mock):
+        """Best Expected Improvement corresponds to the best prediction
+        , as less than r_min proposals made."""
```
Comment (micahjsmith) [inline tests/btb/tuning/test_gp.py]: Fix `,` beginning the line

File: tests/btb/tuning/test_gp.py [hunk 12]
```diff
@@ -186,13 +186,33 @@ class TestGPEi(TestCase):
...
+        # Set-up
+        tuner = GPEi(tuple(), r_minimum=2)
+        tuner.X = np.array([[]])
```
Comment (micahjsmith) [inline tests/btb/tuning/test_gp.py]: are you trying to initialize an empty array of shape=(1, 0)? why not be explicit
```python
tuner.X = np.empty(shape=(1, 0))
```

File: tests/btb/tuning/test_gp.py [hunk 13]
```diff
@@ -186,13 +186,33 @@ class TestGPEi(TestCase):
...
+        tuner.y = np.array([])
+
+        # Run
+        predictions = np.array([
+            [0.8, 2],
+            [0.9, 1]
+        ])
+        best = tuner._acquire(predictions)
+
+        # assert
```
Comment (micahjsmith) [inline tests/btb/tuning/test_gp.py]:
```python
expected_best = 1
self.assertEqual(best, expected_best)
```

File: tests/btb/tuning/test_gp.py [hunk 14]
```diff
@@ -186,13 +186,33 @@ class TestGPEi(TestCase):
     # METHOD: _acquire(self, predictions)
     # VALIDATE:
     #     * return values according to the formula
+    #@patch('btb.tuning.gp.GP')
```
Comment (micahjsmith) [inline tests/btb/tuning/test_gp.py]: why is this here? delete if not needed

File: tests/btb/tuning/test_gp.py [hunk 15]
```diff
@@ -186,13 +186,33 @@ class TestGPEi(TestCase):
...
+        assert best == 1
+        #gp_mock._acquire.assert_called_once_with(predictions)
 
     def test__acquire_best_ei_eq_best_y(self):
         """Best Expected Improvement corresponds to the best prediction."""
 
         # Set-up
         tuner = GPEi(tuple(), r_minimum=2)
         tuner.y = np.array([0.5, 0.6, 0.7])
+        tuner.X = np.array([[1,1],[1,1]])
```
Comment (micahjsmith) [inline tests/btb/tuning/test_gp.py]:
```python
np.ones(shape=(2, 2))
```

File: tests/btb/tuning/test_tuner.py [hunk 16]
```diff
@@ -135,6 +135,24 @@ def test__create_candidates_every_point_used(self):
...
+        tuner = BaseTuner(tunables, gridding=5)
+
+        # Insert 25 part_vecs into X
+        tuner.X = np.array(list(product(*tuner._grid_axes)))
+
+        # Run
+        candidates = tuner._create_candidates(5)
+
+        # Assert
+        assert candidates is None
```
Comment (micahjsmith) [inline tests/btb/tuning/test_tuner.py]:
```
self.assertIsNone(candidates)
```

File: tests/btb/recommendation/test_recommender.py [hunk 17]
```diff
@@ -23,6 +23,7 @@ def test___init__(self):
         np.testing.assert_array_equal(recommender.dpp_matrix, self.dpp_matrix)
         # assert dpp_vector has same number of entries as pipelines
         assert recommender.dpp_vector.shape[0] == self.dpp_matrix.shape[1]
+        assert len(recommender.tried_pipelines) == 0
```
Comment (micahjsmith) [inline tests/btb/recommendation/test_recommender.py]:
```python
self.assertEqual(len(recommender.tried_pipelines), 0)
```

**Comment by pvk-developer** [comment]
Due to major rework and that this pull request refers to an older version of BTB we are closing it. We can still see the changes if needed for future inspiration.
